specs/src/tx_client.md

```diff
@@ -0,0 +1,253 @@
+   - Call gas estimation service to determine gas usage and price
+   - If simulation fails due to a sequence mismatch:
+        - Update local sequence to expected.
+        - Resign and retry gas estimation.
+3. **Fee Setting**: Fee is computed as `ceil(minGasPrice * gasLimit)` unless explicitly set by the user
+4. **Broadcast**: Transaction is encoded and submitted via gRPC:
+    - **Single-endpoint mode**: Send to the default RPC endpoint.
+    - **Multi-endpoint mode**: Broadcast to all endpoints, accept the first success and cancel the rest; if all fail, return the first error.
+5. **Broadcast Error Handling**: Common to both **single** and **multi-endpoint** modes.
+    - Sequence mismatch during CheckTx (mempool validation):
+        - Retry with corrected sequence until success or another failure.
+        - Update the signer’s local sequence to the expected value returned by the node.
+        - **Reasoning**: Without updating to the chain’s expected sequence, the client’s local signer falls behind and stalls. Once stalled, no new transactions can be submitted until the process is manually reset.
+    - Other rejections:
+       - Tendermint/CometBFT does not throw mempool errors for application rejections; they are set in the transaction response.
+       - The client must parse the response code:
+            - 0(`abci.CodeTypeOK`) - accepted into mempool.
+            - Non-zero - tx was rejected; populate and return `BroadcastTxError`.
+
+                 ```go
+                  type BroadcastTxError struct {
+                  TxHash   string // Transaction hash
+                  Code     uint32 // Error code from node
+                  ErrorLog string // Detailed error message
+                }
+                ```
+
+6. **Tracker Entry**: On success, record (`signer`, `sequence`, `txBytes`, `timestamp`) in TxTracker. Entries older than 10 minutes are pruned automatically.
+7. **Sequence Increment**: After successful broadcast, increment the signer’s local sequence by calling **signer.SetSequence()**
+
+### Confirmation
+
+After broadcast, the TxClient continuously polls the chain for transaction status until resolution. This is the tx confrimation life cycle:
+
+- **Pending**: Transaction in the mempool but not yet included, continue polling
+
+- **Committed**: Transaction included in a block.
+  - If execution succeeded (response code = `abci.CodeTypeOK`): return success
+  - If execution failed (any other response code):
+    - Populate `ExecutionError` and return it to the user
+
+    ```go
+    type ExecutionError struct {
+        TxHash   string // Transaction hash
+        Code     uint32 // Execution result code
+        ErrorLog string // Execution error details
+    }
+    ```
+
+- **Rejected**: Transaction was explicitly refused by the node during `ReCheck()` (after each block commit).
+  - Retrieve tx sequence from `TxTracker`
+  - Roll back the sequence to the retrieved tx sequence
```

Comment:
should this also happen if we get error related to sequence? I saw `TxRejected` errors with `WrongSequence`, does it still make sense to revert the local sequence in such case?

Reply:
maybe it should be re-signed and re-submitted on sequence errors?

Reply:
> I saw TxRejected errors with WrongSequence, does it still make sense to revert the local sequence in such case?

This should not really happen now that we have sequential submissions where you basically wait for tx confirmation before you submit another one with the same address. 

Sequence error rejections happen because you submitted a bunch of txs from the same address in parallel and one of them got rejected for a different reason which got all the subsequent ones to have sequence mismatch.

Reply:
wondering if it should still be outlined there, it can still happen if someone shares the same private key between `TxClient` and wallet or so